File: C++/longest_subarray_sum_k.cpp
//TC O(nlogn) SC O(n)
// works for positive and negative numbers

#include <bits/stdc++.h>
using namespace std;

int main(){
    int n;
    cin >> n;
    vector<int> nums(n);
    for(int i = 0; i < n; i++){
        cin >> nums[i];
    }
    long long k;
    cin >> k;

    map<long long, int> presumMP;
    long long sum = 0;
    int max_len = 0;
    for(int i = 0; i<n; i++){
        sum += nums[i];
        if(sum == k){
            max_len = max(max_len, i+1);
        }
        long long rem = sum-k;
        if(presumMP.find(rem) != presumMP.end()){
            int len = i - presumMP[rem];
            max_len = max(max_len, len);
        }
        if(presumMP.find(sum) == presumMP.end()){
            presumMP[sum] = i;
        }
    }
    cout << max_len << endl;
}

File: Python/longest_subarray_sum_k.py
# TC O(2n) (worst case TC) SC O(1)

n = int(input())
nums = []
for _ in input().split():
    nums.append(int(_))

k = int(input())
left = 0
right = 0
sum = nums[0] if n > 0 else 0
max_length = 0
while right < n:
    while sum > k and left <= right:
        sum -= nums[left]
        left += 1
    if sum == k:
        max_length = max(max_length, right - left + 1)
    right += 1        
    if right < n:
        sum += nums[right]
print(max_length)
    